evm_loader/ERC20/test/test_erc20.py

```diff
@@ -14,6 +16,163 @@
+        print("erc20_id:", self.contract_account)
+        print("erc20_code:", self.contract_code_account)
+        if ethereum_id is not None:
+            self.ethereum_id = ethereum_id
+            print("erc20_id_ethereum:", self.ethereum_id.hex())
+        self.caller_ether = None
+        self.neon_evm_client = None
+
+    def set_caller(self, caller_ether):
+        self.caller_ether = caller_ether
+
+    def set_neon_evm_client(self, neon_evm_client):
+        self.neon_evm_client = neon_evm_client
+
+    def balance_ext(self):
+        ethereum_transaction = EthereumTransaction(self.caller_ether, self.contract_account, self.contract_code_account)
+        ethereum_transaction.prepare(abi.function_signature_to_4byte_selector('balance_ext()'))
+        result = self.neon_evm_client.send_ethereum_trx_single(ethereum_transaction)
+        print(result)
+        src_data = result['result']['meta']['innerInstructions'][-1]['instructions'][-1]['data']
+        data = base58.b58decode(src_data)
+        instruction = data[0]
+        assert(instruction == 6)  # 6 means OnReturn
+        assert(data[1] < 0xd0)  # less 0xd0 - success
+        value = data[2:]
+        balance_address = base58.b58encode(value)
+        return balance_address
+
+    def mint_id(self):
+        ethereum_transaction = EthereumTransaction(self.caller_ether, self.contract_account, self.contract_code_account)
+        ethereum_transaction.prepare(abi.function_signature_to_4byte_selector('mint_id()'))
+        result = self.neon_evm_client.send_ethereum_trx_single(ethereum_transaction)
+        print(result)
+        src_data = result['result']['meta']['innerInstructions'][-1]['instructions'][-1]['data']
+        data = base58.b58decode(src_data)
+        instruction = data[0]
+        assert(instruction == 6)  # 6 means OnReturn
+        assert(data[1] < 0xd0)  # less 0xd0 - success
+        value = data[2:]
+        mint_id = base58.b58encode(value)
+        return mint_id
+
+    def balance(self):
+        ethereum_transaction = EthereumTransaction(self.caller_ether, self.contract_account, self.contract_code_account)
+        func_name = abi.function_signature_to_4byte_selector('balanceOf(address)')
+        trx_data = func_name + bytes.fromhex(
+            str("%024x" % 0) +
+            ethereum_transaction.ether_caller.hex()
+        )
+        ethereum_transaction.prepare(trx_data)
+        result = self.neon_evm_client.send_ethereum_trx(ethereum_transaction)
+        print(result)
+        src_data = result['result']['meta']['innerInstructions'][-1]['instructions'][-1]['data']
+        data = base58.b58decode(src_data)
+        instruction = data[0]
+        assert(instruction == 6)  # 6 means OnReturn
+        assert(data[1] < 0xd0)  # less 0xd0 - success
+        value = data[2:]
+        balance = int.from_bytes(value, "big")
+        return balance
+
+    def deposit(self, payer, receiver_erc20, amount, balance_erc20, mint_id, signer):
```

Comment:
It's better to pass transaction `sender` as the argument to the function (instead of passing it through the `set_caller` function). In this case, the sender will be shown in the test clearly.

You need to pay attention that there are two classes of functions:
1 - where you need to form an Ethereum transaction since the contract verifies the sender's signature (deposit, withdraw and transfer); 
2 - the method can be called on behalf of any sender (but we need to create the correct contract call) - balance, balance_ext, mint_id.

Reply:
Fixed